numpy/core/tests/test_dlpack.py

```diff
@@ -110,3 +110,21 @@ def test_readonly(self):
         x.flags.writeable = False
         with pytest.raises(TypeError):
             x.__dlpack__()
+
+    def test_ndim0(self):
+        x = np.array(1.0)
+        y = np._from_dlpack(x)
+        assert_array_equal(x, y)
+
+    def test_from_dlpack_twice(self):
+        x = np.arange(5)
+        capsule = x.__dlpack__()
+        y = np._from_dlpack(capsule, _testing=True)
+        with pytest.raises(ValueError, match=r"PyCapsule_GetPointer called "
+                                             r"with incorrect name"):
+            np._from_dlpack(capsule, _testing=True)
```

Comment:
I don't think I like a "private" kwarg here, and I don't quite understand why it is necessary.  It seems to do nothing except an INCREF and that incref seems odd?!

The allocation change looks nice on first sight though!

Reply:
> I don't think I like a "private" kwarg here, and I don't quite understand why it is necessary.

To check that consumed capsules raise and that the `array_dlpack_deleter` skips deleting a consumed capsule. But it doesn't have to be tested explicitly so I can revert that part. I added it to address "testing calling _from_dlpack twice" semantics.

Reply:
I have reverted. I am not yet aware of how NumPy uses `valgrind` and `pytest-leaks` in its tests/CI so I will add the tests back later if needed. Now the PR is stripped down to testing with `ndims=0` and memory optimization. Let me know if that looks good!

Reply:
> I have reverted. I am not yet aware of how NumPy uses valgrind and pytest-leaks

It doesn't by default, there is a bit of docs on it on the dev guide, but I am  not expecting you to do it.  It is more of a: don't worry about it _too_ much, because we would probably catch it eventually.

I see you removed the test again?  But the test seems right, the capsule will be renamed so an error should be given, how come you can't keep the test?

Reply:
> But the test seems right, the capsule will be renamed so an error should be given, how come you can't keep the test?

`_from_dlpack` doesn't accept capsules; even the one named `"dltensor"`. (it needs an object that has `__dlpack__` attribute). So, I had to remove the test (I added the `_testing` kwargs to enable passing capsule so I could write the test.).

Reply:
Well, you could create a dummy class with a `__dlpack__` dunder to do it.  But, we can just focus on the other change in this PR, and then create another (or not, not like it is super vital).

Reply:
> you could create a dummy class with a `__dlpack__` dunder to do it

Good idea! Thanks!



> But, we can just focus on the other change in this PR, and then create another

OK, sounds good!